crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -325,6 +341,33 @@ impl<T: Send + Sync + Clone, W: Send + Sync> OnchainOrderParser<T, W> {
     }
 }
 
+async fn get_block_numbers_of_events(
+    web3: &Web3,
+    events: &[EthContractEvent<ContractEvent>],
+) -> Result<HashMap<u64, u32>> {
+    let mut event_block_numbers: Vec<u64> = events
+        .iter()
+        .map(|EthContractEvent { meta, .. }| {
+            let meta = match meta {
+                Some(meta) => meta,
+                None => return Err(anyhow!("event without metadata")),
+            };
+            Ok(meta.block_number)
+        })
+        .collect::<Result<Vec<u64>>>()?;
+    event_block_numbers.dedup();
+    let futures = event_block_numbers
+        .into_iter()
+        .map(|block_number| async move {
+            let timestamp =
+                timestamp_of_block_in_seconds(web3, U64::from(block_number).into()).await?;
+            Ok((block_number, timestamp))
+        });
+    let block_number_timestamp_pair: Vec<anyhow::Result<(u64, u32)>> =
+        stream::iter(futures).buffer_unordered(10).collect().await;
```

Comment:
You shouldn't need to `buffer_unordered` here, the `BufferedTransport` implementation we have already does buffering with configurable amounts.

I think it would be fine to just `join_all`.

Reply:
It still has value IMO. The number of events into the function is unbounded. Creating 10000 futures at the same time still uses a lot of memory. I feel we should always bound these kind of tasks and not use join_all if there is no obund.

Reply:
Good point.